app/controllers/studies_controller.rb

```diff
@@ -1223,19 +1223,18 @@ def process_workspace_bucket_files(files)
 
     # next update map of existing files to determine what can be grouped together in a directory listing
     @file_extension_map = DirectoryListing.create_extension_map(files, @file_extension_map)
-
     files.each do |file|
       # check first if file type is in file map in a group larger than 10 (or 20 for text files)
-      file_extension = DirectoryListing.file_extension(file.name)
+      file_type = DirectoryListing.file_type_from_extension(file.name)
       directory_name = DirectoryListing.get_folder_name(file.name)
-      if @file_extension_map.has_key?(directory_name) && !@file_extension_map[directory_name][file_extension].nil? &&
-          @file_extension_map[directory_name][file_extension] >= DirectoryListing::MIN_SIZE
-        process_directory_listing_file(file, file_extension)
+      if @file_extension_map.has_key?(directory_name) && !@file_extension_map.dig(directory_name, file_type).nil? &&
+          @file_extension_map.dig(directory_name, file_type) >= DirectoryListing::MIN_SIZE
+        process_directory_listing_file(file, file_type)
       else
-        # we are now dealing with singleton files or fastqs, so process accordingly (making sure to ignore directories)
-        if DirectoryListing::PRIMARY_DATA_TYPES.any? {|ext| file_extension.include?(ext)} && !file.name.end_with?('/')
-          # process fastq file into appropriate directory listing
-          process_directory_listing_file(file, 'fastq')
+        # we are now dealing with singleton files or sequence data, so process accordingly (making sure to ignore directories)
+        if DirectoryListing::PRIMARY_DATA_TYPES.any? {|ext| file_type.include?(ext)} && !file.name.end_with?('/')
+          # process sequence data file into appropriate directory listing
+          process_directory_listing_file(file, file_type)
         else
```

Comment:
This service and the one above seem to share a lot of code.  Is it worth pulling some of this into the Directory Listing model?  Or maybe a DirectoryListingService?  (or maybe even a more generic StudyService)

Reply:
What I'd actually like to do is this: https://broadworkbench.atlassian.net/browse/SCP-3121

Basically, refactor all code that is "sync" related into a service and DRY up everything else, much like we did with the FileParseService and the various viz-related services.  It would improve both test coverage and maintainability.